### PR TITLE
build: use the unified pickle loader to handle more edge cases

### LINKED ISSUE #11051: meson crash: mesonbuild.mesonlib is not a package.
The command ` meson configure . -Doptimization=0` crashes with a python stack trace:
```
Traceback (most recent call last):
  File "/Users/john/.local/share/virtualenvs/etc-vauAPygS/lib/python3.10/site-packages/mesonbuild/mesonmain.py", line 194, in run
    return options.run_func(options)
  File "/Users/john/.local/share/virtualenvs/etc-vauAPygS/lib/python3.10/site-packages/mesonbuild/mconf.py", line 300, in run
    c = Conf(builddir)
  File "/Users/john/.local/share/virtualenvs/etc-vauAPygS/lib/python3.10/site-packages/mesonbuild/mconf.py", line 72, in __init__
    self.build = build.load(self.build_dir)
  File "/Users/john/.local/share/virtualenvs/etc-vauAPygS/lib/python3.10/site-packages/mesonbuild/build.py", line 2907, in load
    obj = pickle.load(f)
ModuleNotFoundError: No module named 'mesonbuild.mesonlib.universal'; 'mesonbuild.mesonlib' is not a package

ERROR: Unhandled python exception

    This is a Meson bug and should be reported!
```

Source is [cairo](https://gitlab.freedesktop.org/cairo/cairo/) at [commit 6dc5df6](https://gitlab.freedesktop.org/cairo/cairo/-/commit/c6dc5df61264be1bec2e622355b97c6c45d236fc).

Native compilation on macOS 13.1beta with Xcode 14.1's toolchain. Python is a locally-compiled 3.10 in a pyenv-controlled virtenv, and meson-0.64.0 was just pulled from pypi with pip3.

### PR DESCRIPTION
**Pull request by eli-schwartz**
We have divergent implementations of loading a pickled *.dat file. The Build class loader has a better error message, and also checks if the file exists. But the generic loader handles TypeError and ModuleNotFoundError. Merge the implementations, and use it for Build as well.

Fixes #11051

### CODE REVIEW

File: mesonbuild/utils/universal.py [hunk 1]
```diff
@@ -2289,16 +2289,21 @@ def pickle_load(filename: str, object_name: str, object_type: T.Type) -> T.Any:
...
+            f'meson setup {build_dir} --wipe')
     if not isinstance(obj, object_type):
         raise MesonException(load_fail_msg)
     from ..coredata import version as coredata_version
     from ..coredata import major_versions_differ, MesonVersionMismatchException
-    if major_versions_differ(obj.version, coredata_version):
-        raise MesonVersionMismatchException(obj.version, coredata_version)
+    version = getattr(obj, 'version', None)
+    if version is None:
+        version = obj.environment.coredata.version
```
Comment (dcbaker) [inline mesonbuild/utils/universal.py:2304]: Isn't this just:
```python
version = getattr(obj, 'version', obj.environment.coredata.version)
```

**Comment by dcbaker** [comment]
Other than that, this looks good to me

File: mesonbuild/utils/universal.py (hunk 1 above)
Comment (eli-schwartz) [reply mesonbuild/utils/universal.py:2304]: That's an error if it doesn't have an environment attribute.

Coredata and installdata only have a version attribute. build data only has an environment attribute.

We could add a version attribute to `build.Build` but that doesn't help to detect the problem of people loading build directories created by versions of Meson before 0.64.1

Comment (dcbaker) [reply mesonbuild/utils/universal.py:2304]: Oh, right, because it's evaluated eagerly. Sigh

**Review by dcbaker — APPROVED** [review]

**Pull request merged**